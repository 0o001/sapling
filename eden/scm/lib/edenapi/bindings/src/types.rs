/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! These types have a representation that is visible to C / C++, and are either
//! allocated in C / C++ and must be freed accordingly, or do not need to be freed
//! at all (no heap allocations).

use anyhow::Error;
use anyhow::Result;
use edenapi_types::metadata::ContentId as ApiContentId;
use edenapi_types::metadata::Sha1 as ApiSha1;
use edenapi_types::metadata::Sha256 as ApiSha256;
use edenapi_types::FileType as ApiFileType;
use edenapi_types::TreeAttributes as ApiTreeAttributes;
use libc::size_t;
use types::HgId as ApiHgId;
use types::Key as ApiKey;
use types::Parents as ApiParents;
use types::RepoPathBuf;

use crate::ptr_len_to_slice;

#[repr(C)]
pub struct Key {
    path: *const u8,
    path_len: size_t,
    hgid: [u8; 20],
}

// Conversion takes a reference to Key, leaving the caller
// responsible for deallocation of the string.
impl TryFrom<&Key> for ApiKey {
    type Error = Error;
    fn try_from(v: &Key) -> Result<Self> {
        let path = unsafe { ptr_len_to_slice(v.path, v.path_len)? };
        // to_vec copies the slice, so the ApiKey can be
        // deallocated in Rust-land.
        let path = RepoPathBuf::from_utf8(path.to_vec())?;
        Ok(Self {
            path,
            hgid: ApiHgId::from_byte_array(v.hgid),
        })
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub enum FileType {
    Regular,
    Executable,
    Symlink,
}

impl From<ApiFileType> for FileType {
    fn from(v: ApiFileType) -> Self {
        match v {
            ApiFileType::Regular => FileType::Regular,
            ApiFileType::Executable => FileType::Executable,
            ApiFileType::Symlink => FileType::Symlink,
        }
    }
}

#[repr(C, u8)]
#[derive(Copy, Clone)]
pub enum Parents {
    None,
    One([u8; 20]),
    Two([u8; 20], [u8; 20]),
}

impl From<ApiParents> for Parents {
    fn from(v: ApiParents) -> Self {
        match v {
            ApiParents::None => Parents::None,
            ApiParents::One(id1) => Parents::One(id1.into_byte_array()),
            ApiParents::Two(id1, id2) => Parents::Two(id1.into_byte_array(), id2.into_byte_array()),
        }
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct HgId(pub [u8; 20]);

impl From<ApiHgId> for HgId {
    fn from(v: ApiHgId) -> Self {
        HgId(v.into_byte_array())
    }
}

impl From<[u8; 20]> for HgId {
    fn from(v: [u8; 20]) -> Self {
        HgId(v)
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct ContentId(pub [u8; 32]);

impl From<ApiContentId> for ContentId {
    fn from(v: ApiContentId) -> Self {
        ContentId(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha1(pub [u8; 20]);

impl From<ApiSha1> for Sha1 {
    fn from(v: ApiSha1) -> Self {
        Sha1(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha256(pub [u8; 32]);

impl From<ApiSha256> for Sha256 {
    fn from(v: ApiSha256) -> Self {
        Sha256(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct TreeAttributes {
    manifest_blob: bool,
    parents: bool,
    child_metadata: bool,
}

impl From<TreeAttributes> for ApiTreeAttributes {
    fn from(v: TreeAttributes) -> Self {
        ApiTreeAttributes {
            manifest_blob: v.manifest_blob,
            parents: v.parents,
            child_metadata: v.child_metadata,
        }
    }
}
