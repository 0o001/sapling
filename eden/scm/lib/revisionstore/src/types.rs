/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use anyhow::Result;
use bytes::Bytes;
use serde_derive::{Deserialize, Serialize};
use sha2::Digest;

#[cfg(any(test, feature = "for-tests"))]
use rand::Rng;

use types::{Key, Sha256};

/// Kind of content hash stored in the LFS pointer. Adding new types is acceptable, re-ordering or
/// removal is forbidden.
#[derive(
    Serialize,
    Deserialize,
    PartialEq,
    Eq,
    Debug,
    Hash,
    Ord,
    PartialOrd,
    Clone
)]
pub enum ContentHash {
    Sha256(Sha256),
}

#[derive(Clone, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
pub enum StoreKey {
    HgId(Key),
    /// The Key is a temporary workaround to being able to fallback from the LFS protocol to the
    /// non-LFS protocol. Do not depend on it as it will be removed.
    Content(ContentHash, Option<Key>),
}

impl ContentHash {
    pub fn sha256(data: &Bytes) -> Result<Self> {
        let mut hash = sha2::Sha256::new();
        hash.input(data);

        let bytes: [u8; Sha256::len()] = hash.result().into();
        Ok(ContentHash::Sha256(Sha256::from_slice(&bytes)?))
    }

    pub fn unwrap_sha256(self) -> Sha256 {
        match self {
            ContentHash::Sha256(hash) => hash,
        }
    }
}

impl StoreKey {
    pub fn hgid(key: Key) -> Self {
        StoreKey::HgId(key)
    }

    pub fn content(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }
}

impl From<Key> for StoreKey {
    fn from(k: Key) -> Self {
        StoreKey::HgId(k)
    }
}

impl<'a> From<&'a Key> for StoreKey {
    fn from(k: &'a Key) -> Self {
        StoreKey::HgId(k.clone())
    }
}

impl From<ContentHash> for StoreKey {
    fn from(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }
}

impl<'a> From<&'a ContentHash> for StoreKey {
    fn from(hash: &'a ContentHash) -> Self {
        StoreKey::Content(hash.clone(), None)
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl quickcheck::Arbitrary for ContentHash {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        ContentHash::Sha256(Sha256::arbitrary(g))
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl quickcheck::Arbitrary for StoreKey {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        if g.gen() {
            StoreKey::HgId(Key::arbitrary(g))
        } else {
            StoreKey::Content(ContentHash::arbitrary(g), Option::arbitrary(g))
        }
    }
}
