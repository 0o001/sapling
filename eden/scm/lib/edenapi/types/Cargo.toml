# @generated by autocargo from //eden/scm/lib/edenapi/types:edenapi_types
[package]
name = "edenapi_types"
version = "0.1.0"
edition = "2018"

[dependencies]
anyhow = "1.0"
bytes = { version = "1.0", features = ["serde"] }
dag-types = { path = "../../dag/dag-types", features = ["for-tests"] }
faster-hex = "0.4"
quickcheck = "1.0"
rand = { version = "0.8", features = ["small_rng"] }
revisionstore_types = { path = "../../revisionstore/types" }
serde = { version = "1.0.126", features = ["derive", "rc"] }
serde_bytes = "0.11"
serde_derive = "1.0"
serde_json = { version = "1.0", features = ["float_roundtrip"] }
thiserror = "1.0.29"
types = { path = "../../types" }

[dev-dependencies]
paste = "1.0"
quickcheck_macros = "1.0"
serde_cbor = "0.11"

[features]
default = ["for-tests"]
for-tests = []
