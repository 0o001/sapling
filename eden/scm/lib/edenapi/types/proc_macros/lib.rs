/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

extern crate proc_macro;

use std::collections::HashSet;

use proc_macro2::TokenStream;
use quote::format_ident;
use quote::quote;
use syn::spanned::Spanned;
use syn::*;

const ID: &str = "id";

/// Derive a default implementation for a wire type for this type
// TODO: Future improvements
// - Support fields that do not implement Default on Api obj
//    - add "no_default" attribute to field. Wire impl will wrap it in Option,
//      and fail when deserializing if it's not present
// - Support generics in type
//    - Might be possible to make it work with some adaptation, at least for
//      simple "wrapper" types that just have one generic T.
#[proc_macro_attribute]
pub fn auto_wire(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let _ = parse_macro_input!(attr as syn::parse::Nothing);
    let mut item = parse_macro_input!(item as Item);

    let result = match &mut item {
        Item::Struct(ref mut item) => get_wire_struct(item),
        Item::Enum(ref mut item) => get_wire_enum(item),
        item => Err(Error::new(item.span(), "Only struct or enum is supported")),
    };
    match result {
        Ok(wire_item) => quote! {
            #item

            #wire_item
        },
        Err(e) => e.to_compile_error(),
    }
    .into()
}

struct Parenthesized<T>(T);
impl<T> parse::Parse for Parenthesized<T>
where
    T: std::str::FromStr,
    T::Err: std::fmt::Display,
{
    fn parse(input: parse::ParseStream) -> Result<Self> {
        let content;
        parenthesized!(content in input);
        Ok(Self(content.parse::<LitInt>()?.base10_parse::<T>()?))
    }
}

fn arbitrary_impl(wire_ident: &Ident, generics: &Generics) -> TokenStream {
    quote! {
        #[cfg(any(test, feature = "for-tests"))]
        impl #generics quickcheck::Arbitrary for #wire_ident #generics {
            fn arbitrary(g: &mut quickcheck::Gen) -> Self {
                use crate::ToWire;
                <Self as crate::ToApi>::Api::arbitrary(g).to_wire()
            }
        }
    }
}

fn extract_id(
    attrs: Vec<Attribute>,
    spanned: &impl Spanned,
    ids: &mut HashSet<u16>,
) -> Result<(u16, Vec<Attribute>)> {
    let (id, other_attrs): (Vec<_>, Vec<_>) =
        attrs.into_iter().partition(|attr| attr.path.is_ident(ID));
    if id.len() != 1 {
        return Err(Error::new(
            spanned.span(),
            "Must have exactly one attribute 'id'",
        ));
    }
    let id = id
        .into_iter()
        .next()
        // never panics because of if above
        .unwrap();
    let id = parse2::<Parenthesized<u16>>(id.tokens)?.0;
    if !ids.insert(id) {
        return Err(Error::new(
            spanned.span(),
            "'id' attribute should be unique across all fields/variants",
        ));
    }
    Ok((id, other_attrs))
}

fn remove_id(attrs: &mut Vec<Attribute>) {
    *attrs = std::mem::take(attrs)
        .into_iter()
        .filter(|attr| !attr.path.is_ident(ID))
        .collect();
}

fn get_wire_struct(original: &mut ItemStruct) -> Result<TokenStream> {
    let mut item = original.clone();
    let ident = item.ident.clone();
    let wire_ident = format_ident!("Wire{}", ident);
    item.ident = wire_ident.clone();
    item.attrs = vec![
        parse_quote!(#[derive(Default, Debug, serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq)]),
    ];

    let mut fields = vec![];

    let mut ids = HashSet::new();
    match &mut item.fields {
        Fields::Named(ref mut fs) => fs.named.iter_mut().try_for_each(|ref mut field| {
            fields.push(field.ident.clone().unwrap());
            let (id, other_attrs) = extract_id(std::mem::take(&mut field.attrs), &field, &mut ids)?;
            field.attrs = other_attrs;
            let ty = &field.ty;
            field.ty = parse_quote!( <#ty as crate::ToWire>::Wire );
            let name = format!("{}", id);
            field.attrs.push(
                parse_quote!( #[serde(rename=#name, default, skip_serializing_if="crate::wire::is_default")] ),
            );
            Result::Ok(())
        })?,
        _ => {
            return Err(Error::new(
                item.fields.span(),
                "Only structs with named fields supported",
            ));
        }
    }

    // remove id() attribute from original struct
    match &mut original.fields {
        Fields::Named(ref mut fs) => fs.named.iter_mut().for_each(|ref mut field| {
            remove_id(&mut field.attrs);
        }),
        _ => unreachable!(),
    }

    let fields_to_wire = fields
        .iter()
        .map(|name| quote! { #name: self.#name.to_wire() });

    let generics = &original.generics;

    let to_wire_impl = quote! {
        impl #generics crate::ToWire for #ident #generics {
            type Wire = #wire_ident #generics;

            fn to_wire(self) -> Self::Wire {
                Self::Wire {
                    #( #fields_to_wire ),*
                }
            }
        }
    };

    let fields_to_api = fields
        .iter()
        .map(|name| quote! { #name: self.#name.to_api()? });

    let to_api_impl = quote! {
        impl #generics crate::ToApi for #wire_ident #generics {
            type Api = #ident #generics;
            type Error = crate::WireToApiConversionError;

            fn to_api(self) -> Result<Self::Api, Self::Error> {
                Ok(Self::Api {
                    #( #fields_to_api ),*
                })
            }
        }
    };

    let arbitrary_impl = arbitrary_impl(&wire_ident, generics);

    Ok(quote! {
        #item
        #to_wire_impl
        #to_api_impl
        #arbitrary_impl
    })
}

fn get_wire_enum(original: &mut ItemEnum) -> Result<TokenStream> {
    // original is the original Enum
    // item is the counterpart WireEnum
    let mut item = original.clone();
    let ident = item.ident.clone();
    let wire_ident = format_ident!("Wire{}", ident);
    item.ident = wire_ident.clone();
    item.attrs = vec![
        parse_quote!(#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq)]),
    ];

    let mut variants = vec![];
    let mut ids = HashSet::new();
    item.variants.iter_mut().try_for_each(|ref mut variant| {
        let (id, other_attrs) = extract_id(std::mem::take(&mut variant.attrs), &variant, &mut ids)?;
        variant.attrs = other_attrs;
        if id == 0 {
            return Err(Error::new(variant.span(), "Variant id can't be 0"));
        }
        let name = format!("{}", id);
        variant.attrs.push(parse_quote!( #[serde(rename=#name)]));
        let unit = match &mut variant.fields {
            Fields::Unit => true,
            Fields::Unnamed(fields) if fields.unnamed.len() == 1 => {
                let field = fields.unnamed.first_mut().unwrap();
                let ty = &field.ty;
                field.ty = parse_quote!( <#ty as crate::ToWire>::Wire );
                false
            }
            _ => {
                return Err(Error::new(
                    variant.fields.span(),
                    "Only unit variants or with a single field supported",
                ));
            }
        };
        variants.push((variant.ident.clone(), unit));
        Ok(())
    })?;
    item.variants.push_value(parse_quote!(
        #[serde(other, rename = "0")]
        UnknownVariant
    ));

    // remove id() attribute from original enum
    original.variants.iter_mut().for_each(|ref mut variant| {
        remove_id(&mut variant.attrs);
    });

    let variants_to_wire = variants.iter().map(|(name, unit)| {
        if *unit {
            quote! { Self::#name => Self::Wire::#name }
        } else {
            quote! { Self::#name(value) => Self::Wire::#name(value.to_wire()) }
        }
    });

    let generics = &original.generics;

    let to_wire_impl = quote! {
        impl #generics crate::ToWire for #ident #generics {
            type Wire = #wire_ident #generics;

            fn to_wire(self) -> Self::Wire {
                match self {
                    #( #variants_to_wire ),*
                }
            }
        }
    };

    let variants_to_api = variants.iter().map(|(name, unit)| {
        if *unit {
            quote! { Self::#name => Ok(Self::Api::#name) }
        } else {
            quote! { Self::#name(value) => Ok(Self::Api::#name(value.to_api()?)) }
        }
    });

    let to_api_impl = quote! {
        impl #generics crate::ToApi for #wire_ident #generics {
            type Api = #ident #generics;
            type Error = crate::WireToApiConversionError;

            fn to_api(self) -> Result<Self::Api, Self::Error> {
                match self {
                    #( #variants_to_api, )*
                    Self::UnknownVariant => Err(Self::Error::UnrecognizedEnumVariant(stringify!(#wire_ident))),
                }
            }
        }
    };

    let arbitrary_impl = arbitrary_impl(&wire_ident, generics);

    let default_impl = quote! {
        impl #generics Default for #wire_ident #generics {
            fn default() -> Self {
                use crate::ToWire;
                #ident::default().to_wire()
            }
        }
    };

    Ok(quote! {
        #item
        #to_wire_impl
        #to_api_impl
        #arbitrary_impl
        #default_impl
    })
}

#[test]
#[should_panic = "'id' attribute should be unique across all fields/variants"]
fn test_same_ids() {
    let mut input = parse_quote! {
        enum MyEnum {
            #[id(1)]
            A,
            #[id(1)]
            B,
        }
    };
    get_wire_enum(&mut input).unwrap();
}

#[test]
#[should_panic = "Must have exactly one attribute 'id'"]
fn test_two_ids() {
    let mut input = parse_quote! {
        enum MyEnum {
            #[id(1)]
            #[id(1)]
            A,
        }
    };
    get_wire_enum(&mut input).unwrap();
}
