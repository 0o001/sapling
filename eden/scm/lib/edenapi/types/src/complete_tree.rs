/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use serde_derive::{Deserialize, Serialize};

use types::{hgid::HgId, path::RepoPathBuf};

/// Struct reprenting the arguments to a "gettreepack" operation, which
/// is used by Mercurial to prefetch treemanifests. This struct is intended
/// to provide a way to support requests compatible with Mercurial's existing
/// gettreepack wire protocol command.
///
/// In the future, we'd like to migrate away from requesting trees in this way.
/// In general, trees can be requested from the API server using a `TreeRequest`
/// containing the keys of the desired tree nodes.
///
/// In all cases, trees will be returned in a `TreeResponse`.
#[derive(Clone, Debug, Eq, Deserialize, Serialize, PartialEq)]
pub struct CompleteTreeRequest {
    pub rootdir: RepoPathBuf,
    pub mfnodes: Vec<HgId>,
    pub basemfnodes: Vec<HgId>,
    pub depth: Option<usize>,
}

impl CompleteTreeRequest {
    pub fn new(
        rootdir: RepoPathBuf,
        mfnodes: Vec<HgId>,
        basemfnodes: Vec<HgId>,
        depth: Option<usize>,
    ) -> Self {
        Self {
            rootdir,
            mfnodes,
            basemfnodes,
            depth,
        }
    }
}

#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for CompleteTreeRequest {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        Self {
            rootdir: Arbitrary::arbitrary(g),
            mfnodes: Arbitrary::arbitrary(g),
            basemfnodes: Arbitrary::arbitrary(g),
            depth: Arbitrary::arbitrary(g),
        }
    }
}
