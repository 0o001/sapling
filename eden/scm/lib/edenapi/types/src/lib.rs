/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Types shared between the EdenAPI client and server.
//!
//! This crate exists primarily to provide a lightweight place to
//! put types that need to be used by both the client and server.
//! Types that are exclusive used by either the client or server
//! SHOULD NOT be added to this crate.
//!
//! Given that the client and server are each part of different
//! projects (Mercurial and Mononoke, respectively) which have
//! different build processes, putting shared types in their own
//! crate decreases the likelihood of build failures caused by
//! dependencies with complex or esoteric build requirements.
//!
//! Most of the types in this crate are used for data interchange
//! between the client and server. As such, CHANGES TO THE THESE
//! TYPES MAY CAUSE VERSION SKEW, so any changes should proceed
//! with caution.

#![deny(warnings)]

pub mod commit;
pub mod complete_tree;
pub mod file;
pub mod history;
pub mod json;
pub mod metadata;
pub mod tree;
pub mod wire;

pub use crate::commit::{
    CommitLocation, CommitLocationToHash, CommitLocationToHashRequest, CommitRevlogData,
    CommitRevlogDataRequest,
};
pub use crate::complete_tree::CompleteTreeRequest;
pub use crate::file::{FileEntry, FileError, FileRequest};
pub use crate::history::{
    HistoryEntry, HistoryRequest, HistoryResponse, HistoryResponseChunk, WireHistoryEntry,
};
pub use crate::metadata::{
    ContentId, DirectoryMetadata, DirectoryMetadataRequest, FileMetadata, FileMetadataRequest,
    FileType, FsnodeId, Sha1, Sha256,
};
pub use crate::tree::{TreeChildEntry, TreeEntry, TreeError, TreeRequest};
pub use crate::wire::{ToApi, ToWire, WireToApiConversionError};

#[cfg(any(test, feature = "for-tests"))]
use quickcheck::Arbitrary;
use thiserror::Error;

use bytes::Bytes;
use types::{hgid::HgId, key::Key, parents::Parents, path::RepoPathBuf};

#[derive(Debug, Error)]
#[error("Invalid hash: {expected} (expected) != {computed} (computed)")]
pub struct InvalidHgId {
    expected: HgId,
    computed: HgId,
    data: Bytes,
    parents: Parents,
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
#[error("Error fetching key {key:?}: {err}")]
pub struct EdenApiServerError {
    pub err: EdenApiServerErrorKind,
    pub key: Option<Key>,
}

impl EdenApiServerError {
    pub fn new(err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: None,
        }
    }

    pub fn with_key(key: Key, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(key),
        }
    }

    pub fn with_path(path: RepoPathBuf, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(Key {
                path,
                hgid: *HgId::null_id(),
            }),
        }
    }

    pub fn with_hgid(hgid: HgId, err: impl std::fmt::Debug) -> EdenApiServerError {
        EdenApiServerError {
            err: EdenApiServerErrorKind::OpaqueError(format!("{:?}", err)),
            key: Some(Key {
                hgid,
                path: RepoPathBuf::new(),
            }),
        }
    }
}

#[derive(Clone, Debug, Error, Eq, PartialEq)]
pub enum EdenApiServerErrorKind {
    #[error("EdenAPI server returned an error with message: {0}")]
    OpaqueError(String),
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for EdenApiServerError {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        Self {
            err: Arbitrary::arbitrary(g),
            key: Arbitrary::arbitrary(g),
        }
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl Arbitrary for EdenApiServerErrorKind {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        use rand::Rng;
        let variant = g.gen_range(0, 1);
        match variant {
            0 => EdenApiServerErrorKind::OpaqueError(Arbitrary::arbitrary(g)),
            _ => unreachable!(),
        }
    }
}
