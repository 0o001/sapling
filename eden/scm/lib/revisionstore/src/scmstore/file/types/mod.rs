/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod auxdata;
mod lazy_file;
mod store_file;

pub(crate) use self::lazy_file::LazyFile;
pub use self::{attrs::FileAttributes, auxdata::FileAuxData, store_file::StoreFile};
