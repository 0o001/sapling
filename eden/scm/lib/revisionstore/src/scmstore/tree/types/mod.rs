/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod lazy_tree;
mod store_tree;

pub(crate) use self::lazy_tree::LazyTree;
pub use self::{attrs::TreeAttributes, store_tree::StoreTree};
