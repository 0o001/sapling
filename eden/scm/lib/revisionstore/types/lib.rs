/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#![deny(warnings)]

//! revisionstore_types - Data types used by the revisionstore crate

mod datastore;

pub use crate::datastore::Metadata;
