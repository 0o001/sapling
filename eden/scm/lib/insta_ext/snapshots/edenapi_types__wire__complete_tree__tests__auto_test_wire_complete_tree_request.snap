---
source: complete_tree.rs
expression: "WireCompleteTreeRequest::arbitrary(&mut g)"

---
{"0":"erlvio/goxk/vl/qvjafti/vh/xha","1":[[175,25,36,58,55,101,239,1,100,81,21,29,1,4,212,7,211,178,104,186],[230,81,160,133,254,52,10,95,232,130,59,15,255,112,185,80,58,100,255,255],[178,51,90,83,215,163,208,194,73,31,251,99,255,236,0,239,140,183,249,132]],"3":1377840696971470705}
